packages/liveblocks-core/src/room.ts

```diff
@@ -1048,26 +1048,36 @@ function makeStateMachine<
     );
   }
 
-  function apply(
-    ops: HistoryOp<TPresence>[],
+  function applyOps<O extends HistoryOp<TPresence>>(
+    rawOps: readonly O[],
     isLocal: boolean
   ): {
-    reverse: HistoryOp<TPresence>[];
+    // Input Ops can get opIds assigned during application.
+    ops: O[];
+    reverse: O[];
     updates: {
       storageUpdates: Map<string, StorageUpdate>;
       presence: boolean;
     };
   } {
```

Comment:
Apply will now output the (potentially) modified ops array, and won't mutate its input in-place.

Ops originating from undo/redo stacks won't have an `opId` assigned yet.

Before this PR:

```ts
apply(ops, …)
for (const op of ops) {
  // op _will_ have opId set here!
}
```

This led to a few surprises when refactoring this code. This crucial side effect was a bit hidden here and I didn't realize this was happening at first.

After this PR:

```ts
result = applyOps(ops, …)
for (const op of ops) {
  // op.opId might be undefined here! Input array won't be mutated!
}

// Instead, use the _result_ of applyOps()
for (const op of result.ops) {
  // op.opId is set here!
}
```